### ISSUE TITLE
HMR broken in certain cases when using the Storyblok integration

### ISSUE THREAD

**Issue by tony-sull** [withastro/astro#4727]
### What version of `astro` are you using?

1.2.2

### Are you using an SSR adapter? If so, which one?

NOne

### What package manager are you using?

npm

### What operating system are you using?

Mac

### Describe the Bug

HMR doesn't seem to be triggering updates in components used by the Storyblok integration.  Storyblok's API sends down JSON blocks that are mapped back to Astro components based on a `component` string key

I haven't had a chance to dig deeper into this one yet, but a few early thoughts

- I'm wondering if having the the `string -> component` mapping done inside the integration isn't being reflected in the dependency graph in Vite...is this an issue in Astro, the integration, or an API limitation?
- Confirm that styles are being bundled, if it's related to the component HMR there's a chance that scoped styles are being missed as well

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/astro-sdk-demo

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by matthewp** [comment]
Here's a reduced version: https://stackblitz.com/edit/github-ql4awq?file=astro.config.mjs,src%2Fpages%2Findex.astro,src%2Fother%2FOther.astro&on=stackblitz

**Comment by manuelschroederdev** [comment]
Hey @matthewp, should this issue theoretically be fixed with the new 1.2.4 release using Vite 3.1? I've just tested it but the issue remains.

**Comment by matthewp** [comment]
@manuelschroederdev No, @bluwy and I debugged it and there's another Vite bug that needs to be fixed. @bluwy is working on that now.

**Comment by manuelschroederdev** [comment]
Cool, thanks for your hard work!

**Comment by bluwy** [comment]
This one's a rollercoaster by I managed to fix it at https://github.com/vitejs/vite/pull/10144

**Comment by manuelschroederdev** [comment]
Looks like https://github.com/vitejs/vite/pull/10144 is getting rather complicated to solve. 👀 
An interesting discussion that you're having there.

**Comment by bluwy** [comment]
@manuelschroederdev I've chat with patak today and got it resolved. It's planned for Vite 3.2, which the beta would release next week. So it might take a few weeks to reach stable. Is this currently blocking the Storyblok integration? If so, I can try to fast-path it as a patch release as I think it's a safe change.

**Comment by manuelschroederdev** [comment]
Hey @bluwy, that's great news, thanks!
I would not say it's blocking the Storyblok integration per se, but it certainly affects the DX. Having to restart the dev server every time you change a component is not ideal. A patch release would be fantastic and much appreciated, but I can imagine you might have more urgent tasks to work on. 🙂

**Comment by bluwy** [comment]
@manuelschroederdev Vite 3.1.3 is released with the fix! Will update astro to use it.

**Comment by manuelschroederdev** [comment]
Wonderful news, thank you 🙏

**Comment by matthewp** [comment]
@bluwy is there still another fix coming in 3.2? Or is this the only one?
